log4j-api/src/main/java/org/apache/logging/log4j/spi/Provider.java

```diff
@@ -97,127 +217,266 @@ public String getVersions() {
-            LOGGER.error("Unable to create class {} specified in {}", threadContextMap, url.toString(), e);
+            LOGGER.error("Unable to load class {} specified in {}", threadContextMap, url.toString(), e);
         }
         return null;
     }
 
+    /**
+     * Extension point for providers to create a {@link ThreadContextMap}
+     * <p>
+     *     The default implementation:
+     * </p>
+     * <ol>
+     *     <li>calls {@link #loadThreadContextMap},</li>
+     *     <li>if the previous call returns {@code null}, it calls {@link #getThreadContextMap} to instantiate one of
+     *     the internal implementations,</li>
+     *     <li>it returns a no-op map otherwise.</li>
+     * </ol>
+     * @since 2.24.0
+     */
+    protected ThreadContextMap createThreadContextMap() {
+        final Class<? extends ThreadContextMap> threadContextMapClass = loadThreadContextMap();
+        if (threadContextMapClass != null) {
+            try {
+                return LoaderUtil.newInstanceOf(threadContextMapClass);
+            } catch (final Exception e) {
+                LOGGER.error(
+                        "Unable to create instance of class {} specified in {}",
+                        threadContextMapClass.getName(),
+                        getUrl(),
+                        e);
+            }
+        }
+        // Standard Log4j API implementations are internal and can be only specified by name:
+        final String threadContextMap = getThreadContextMap();
+        if (threadContextMap != null) {
+            /*
+             * The constructors are called explicitly to improve GraalVM support.
+             *
+             * The class names of the package-private implementations from version 2.23.1 must be recognized even
+             * if the class is moved.
+             */
+            switch (threadContextMap) {
+                case NO_OP_CONTEXT_MAP:
+                case "org.apache.logging.log4j.spi.NoOpThreadContextMap":
+                    return new NoOpThreadContextMap();
+                case WEB_APP_CONTEXT_MAP:
+                case "org.apache.logging.log4j.spi.DefaultThreadContextMap":
+                    return new DefaultThreadContextMap();
+                case GARBAGE_FREE_CONTEXT_MAP:
+                case "org.apache.logging.log4j.spi.GarbageFreeSortedArrayThreadContextMap":
+                    return new GarbageFreeSortedArrayThreadContextMap();
+                case COPY_ON_WRITE_CONTEXT_MAP:
+                case "org.apache.logging.log4j.spi.CopyOnWriteSortedArrayThreadContextMap":
+                    return new CopyOnWriteSortedArrayThreadContextMap();
+            }
+        }
+        LOGGER.warn("Falling back to {}", NoOpThreadContextMap.class.getName());
+        return new NoOpThreadContextMap();
+    }
+
+    // Used for testing
+    void resetThreadContextMap() {
+        threadContextMapLazy.set(null);
+    }
+
+    /**
+     * @return A lazily initialized thread context map.
```

Comment:
```suggestion
     * @return a new thread context map instance
```
_"A lazily initialized"_ statement is not correct, the returned instance is already initialized.